### PR TITLE
Restructure and optimize `course_file` methods

### PR DESCRIPTION
**Pull request by daniil-berg**
In the `course_file` base class, assign the `$file` as a protected property early on during initialization. Then there is no need to pass it around every method.

Make all methods of the `course_file` sub-classes protected.

Ensure consistent PHPDoc parameters throughout `course_file` and its sub-classes.

Optimize and/or simplify some control flows here and there.

### CODE REVIEW

**Comment by MartinGauk** [comment]
I found the following errors:
```
ERROR: relation "label" does not exist
line 254 of /local/listcoursefiles/classes/course_file.php: call to moodle_database->get_record_sql()
line 125 of /local/listcoursefiles/classes/course_file.php: call to local_listcoursefiles\course_file->is_file_used()
```
and
`Argument 1 passed to local_listcoursefiles\course_file::is_embedded_file_used() must be an object, bool given, called in [dirroot]/local/listcoursefiles/classes/course_file.php on line 260`

File: classes/course_file.php [hunk 1]
```diff
@@ -201,122 +204,107 @@ protected function get_file_download_url($file) {
...
-                            WHERE ctx.id = ?';
-                    $mod = $DB->get_record_sql($sql, [$file->contextid]);
-                    $isused = $this->is_embedded_file_used($mod, 'intro', $file->filename);
+            case 'mod': // Course module.
+                $modname = str_replace('mod_', '', $this->file->component);
+                if ($this->file->filearea === 'intro') {
+                    $sql = "SELECT m.*
+                              FROM {context} ctx
+                              JOIN {course_modules} cm ON cm.id = ctx.instanceid
+                              JOIN {$modname} m ON m.id = cm.instance
```
Comment (MartinGauk) [inline classes/course_file.php:252]: Remember that curly braces have a special meaning in double quoted strings and Moodle needs the curly braces around table names.
To be sure that sql injections are definitely impossible, please check whether `$modname` is the name of an installed plugin. `core_component::get_plugin_list('mod')`

**Review by MartinGauk — CHANGES_REQUESTED** [review]

File: classes/course_file.php (hunk 1 above)
Comment (daniil-berg) [reply classes/course_file.php]: You're right of course. Sorry about that oversight. Should be fixed now. I also added the additional check you requested. Hope it works as expected now.

File: classes/course_file.php [hunk 2]
```diff
@@ -16,6 +16,8 @@
 
 namespace local_listcoursefiles;
 
+use Matrix\Exception;
```
Comment (MartinGauk) [inline classes/course_file.php:19]: ?

Comment (daniil-berg) [reply classes/course_file.php]: Oops.

**Pull request merged**